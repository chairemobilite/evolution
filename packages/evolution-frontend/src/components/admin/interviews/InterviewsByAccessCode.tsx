/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { useTranslation } from 'react-i18next';
import Loadable from 'react-loadable';
import Loader from 'react-spinners/HashLoader';

import { InterviewContext } from '../../../contexts/InterviewContext';
import InputString from 'chaire-lib-frontend/lib/components/input/InputString';
import { Link, useLocation, useParams } from 'react-router';
import { _booleish } from 'chaire-lib-common/lib/utils/LodashExtensions';

const loader = function Loading() {
    return <Loader size={30} color={'#aaaaaa'} loading={true} />;
};

const InterviewsComponent = Loadable({
    // TODO: move this to componentDidMount like in Monitoring.tsx
    loader: () => import('./InterviewSearchList'),
    loading: loader
});

const InterviewsByAccessCode: React.FC = () => {
    const location = useLocation();
    const params = useParams();
    const { t } = useTranslation('admin');

    const urlSearch = new URLSearchParams(location.search);
    const [currentCode, setCurrentCode] = React.useState(params.accessCode);
    const [createNewIfNoData] = React.useState(_booleish(urlSearch.get('autoCreate')) === true);
    const { state, dispatch } = React.useContext(InterviewContext);

    React.useEffect(() => {
        if (params.accessCode !== state.response.accessCode) {
            dispatch({ type: 'update_response', response: { accessCode: params.accessCode } });
        }
    }, [params.accessCode, state.response]);

    return (
        <div className="admin">
            <div className="survey-section__content apptr__form-container">
                <InputString
                    id="accessCodeSearchInput"
                    value={currentCode}
                    onValueUpdated={(newValue) => setCurrentCode(newValue.value)}
                ></InputString>
                <Link to={`/interviews/byCode/${currentCode}`}>
                    <button type="button" className={'survey-section__button button blue small'}>
                        {t('admin:interviewSearch:SearchByCode')}
                    </button>
                </Link>
            </div>
            <InterviewsComponent autoCreateIfNoData={createNewIfNoData} queryData={urlSearch} />
        </div>
    );
};

export default InterviewsByAccessCode;
