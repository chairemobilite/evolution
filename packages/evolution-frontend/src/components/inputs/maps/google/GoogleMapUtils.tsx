/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { FeatureGeocodedProperties } from '../InputMapTypes';
import { point as turfPoint } from '@turf/turf';

export const geojson = (
    latLng: google.maps.LatLng | null | undefined
): GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined => {
    return latLng && latLng.lat() && latLng.lng() ? turfPoint([latLng.lng(), latLng.lat()], {}) : undefined;
};

export const toLatLng = (feature: GeoJSON.Feature<GeoJSON.Point>): google.maps.LatLng => {
    return new google.maps.LatLng(feature.geometry.coordinates[1], feature.geometry.coordinates[0]);
};
