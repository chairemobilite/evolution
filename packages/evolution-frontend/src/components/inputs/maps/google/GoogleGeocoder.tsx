/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { FeatureGeocodedProperties, PlaceGeocodedProperties } from '../InputMapTypes';
import { geojson } from './GoogleMapUtils';
import { point as turfPoint, distance as turfDistance } from '@turf/turf';

// In meters:
const geocodingSearchRadiusMinimum = 500;
const geocodingSearchRadiusMaximum = 8000;

export const geocodeSinglePoint = (
    addressQueryString: string,
    options: { bbox?: [number, number, number, number] }
): Promise<GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined> => {
    return new Promise((resolve, reject) => {
        if (!google) {
            resolve(undefined);
        }
        const geocoder = new google.maps.Geocoder();
        const bounds = options.bbox
            ? new google.maps.LatLngBounds(
                { lat: options.bbox[0], lng: options.bbox[1] },
                { lat: options.bbox[2], lng: options.bbox[3] }
            )
            : undefined;
        geocoder.geocode({ address: addressQueryString, bounds }, (results, status) => {
            if (status === google.maps.GeocoderStatus.OK && results !== null) {
                const feature = geojson(results[0].geometry.location);
                if (feature) {
                    feature.properties.geocodingResultMetadata = {
                        formattedAddress: results[0].formatted_address,
                        precision: results[0].geometry.location_type,
                        types: results[0].types,
                        placeId: results[0].place_id
                    };
                    feature.properties.lastAction = 'geocoding';
                    feature.properties.geocodingQueryString = addressQueryString;
                }
                resolve(feature);
            } else {
                reject(`Geocoding failed for ${addressQueryString}`);
            }
        });
    });
};

export const geocodeMultiplePlaces = (
    geocodingQueryString: string,
    options: { bbox?: [number, number, number, number]; map?: google.maps.Map; language?: string }
): Promise<GeoJSON.Feature<GeoJSON.Point, PlaceGeocodedProperties>[] | undefined> => {
    return new Promise((resolve, reject) => {
        if (!options.map) {
            reject('Undefined map: Cannot use Google geocoder without Google maps');
            return;
        }
        if (!google || !options.bbox) {
            resolve(undefined);
            return;
        }
        const geocoder = new google.maps.places.PlacesService(options.map);
        const bounds = new google.maps.LatLngBounds(
            { lat: options.bbox[0], lng: options.bbox[1] },
            { lat: options.bbox[2], lng: options.bbox[3] }
        );

        const upperBound = turfPoint([options.bbox[0], options.bbox[1]]);
        const lowerBound = turfPoint([options.bbox[2], options.bbox[3]]);
        const viewportRadius = turfDistance(upperBound, lowerBound, { units: 'meters' }) / 2;
        const searchRadius = Math.min(
            Math.max(viewportRadius, geocodingSearchRadiusMinimum),
            geocodingSearchRadiusMaximum
        );

        geocoder.textSearch(
            {
                query: geocodingQueryString,
                location: bounds.getCenter(),
                radius: searchRadius,
                language: options.language
            },
            (results, status) => {
                if (status === google.maps.places.PlacesServiceStatus.OK && results !== null) {
                    const places = results
                        .map(({ geometry, ...googlePlace }) => {
                            const feature = geojson(geometry?.location) as GeoJSON.Feature<
                                GeoJSON.Point,
                                PlaceGeocodedProperties
                            >;
                            if (!feature) {
                                return undefined;
                            }
                            feature.id = googlePlace.place_id;
                            feature.properties.placeData = {
                                ...googlePlace
                            };
                            return feature;
                        })
                        .filter((feature) => feature !== undefined) as GeoJSON.Feature<
                        GeoJSON.Point,
                        PlaceGeocodedProperties
                    >[];

                    resolve(places);
                } else {
                    reject(`Geocoding failed for ${geocodingQueryString}`);
                }
            }
        );
    });
};
