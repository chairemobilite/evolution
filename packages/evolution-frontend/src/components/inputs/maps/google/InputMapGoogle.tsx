/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { GoogleMap, useJsApiLoader, Marker } from '@react-google-maps/api';
import GeoJSON from 'geojson';
import bowser from 'bowser';

import projectConfig from 'chaire-lib-common/lib/config/shared/project.config';
import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { googleMapConfigNew as googleConfig } from '../../../../config/googleMaps.config';
import InputLoading from '../../InputLoading';
import { FeatureGeocodedProperties, MarkerData, InfoWindow } from '../InputMapTypes';
import { geojson, toLatLng } from './GoogleMapUtils';

export interface InputGoogleMapPointProps {
    defaultCenter: { lat: number; lon: number };
    maxGeocodingResultsBounds?: [{ lat: number; lng: number }, { lat: number; lng: number }];
    value?: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties>;
    defaultZoom?: number;
    maxZoom?: number;
    height?: string; // the height of the map container in css units: example: 28rem or 550px
    markers: MarkerData[];
    onValueChange: (feature: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined) => void;
    onMapReady?: (bbox?: [number, number, number, number]) => void;
    onBoundsChanged?: (bbox?: [number, number, number, number]) => void;
    // Change this value when map bounds should be increased to fit markers
    shouldFitBounds?: number;
    infoWindow?: InfoWindow;
    setGeocodingOptions?: (options: { [key: string]: unknown }) => void;
}

const callWithBounds = (
    bounds?: google.maps.LatLngBounds,
    call?: (bbox?: [number, number, number, number]) => void
) => {
    if (!call) {
        return;
    }
    call(
        bounds
            ? [
                bounds.getSouthWest().lat(),
                bounds.getSouthWest().lng(),
                bounds.getNorthEast().lat(),
                bounds.getNorthEast().lng()
            ]
            : undefined
    );
};

const InputMapGoogle: React.FunctionComponent<InputGoogleMapPointProps> = (props: InputGoogleMapPointProps) => {
    const { isLoaded } = useJsApiLoader({
        region: projectConfig.region,
        language: projectConfig.defaultLocale,
        ...googleConfig
    });

    const [map, setMap] = React.useState<google.maps.Map | null>(null);
    const [center, setCenter] = React.useState<{ lat: number; lng: number } | google.maps.LatLng>({
        lat: props.defaultCenter.lat,
        lng: props.defaultCenter.lon
    });
    const [placesInfoWindow, setPlacesInfoWindow] = React.useState<google.maps.InfoWindow | undefined>(undefined);

    const changeValueAndPan = React.useCallback(
        (feature: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties> | undefined) => {
            props.onValueChange(feature);
            if (feature) {
                setCenter({ lat: feature.geometry.coordinates[1], lng: feature.geometry.coordinates[0] });
                (map as google.maps.Map).panTo({
                    lat: feature.geometry.coordinates[1],
                    lng: feature.geometry.coordinates[0]
                });
            }
        },
        [map]
    );

    const onLoad = React.useCallback((map: google.maps.Map) => {
        setMap(map);
        if (props.setGeocodingOptions) props.setGeocodingOptions({ map });
        callWithBounds(map.getBounds(), props.onMapReady);
        setPlacesInfoWindow(
            new google.maps.InfoWindow({
                pixelOffset: new google.maps.Size(0, -40)
            })
        );
    }, []);

    const onUnmount = React.useCallback(() => {
        setMap(null);
        if (props.setGeocodingOptions) props.setGeocodingOptions({ map: undefined });
    }, []);

    const onPositionChange = React.useCallback(
        (e: google.maps.MapMouseEvent, triggerEvent: string) => {
            const geojsonValue = geojson(e.latLng);
            if (geojsonValue) {
                geojsonValue.properties.lastAction = triggerEvent;
                geojsonValue.properties.zoom = (map as google.maps.Map).getZoom();
                geojsonValue.properties.platform = bowser.getParser(window.navigator.userAgent).getPlatformType();
            }
            changeValueAndPan(geojsonValue);
        },
        [map]
    );

    const onZoomChange = () => {
        if (!map) return;
        const currentZoom = (map as google.maps.Map).getZoom();
        if (currentZoom && props.maxZoom && props.maxZoom < currentZoom) {
            (map as google.maps.Map).setZoom(props.maxZoom);
        }
    };

    const onBoundsChanged = () => {
        if (!map) return;
        callWithBounds((map as google.maps.Map).getBounds(), props.onBoundsChanged);
    };

    React.useEffect(() => {
        if (map) {
            (map as google.maps.Map).panTo({
                lat: props.defaultCenter.lat,
                lng: props.defaultCenter.lon
            });
        }
    }, [props.defaultCenter]);

    React.useEffect(() => {
        if (map && props.markers.length >= 1) {
            const maxMapBounds = new google.maps.LatLngBounds();
            if (props.maxGeocodingResultsBounds !== undefined) {
                maxMapBounds.extend(props.maxGeocodingResultsBounds[0]); // lower left
                maxMapBounds.extend(props.maxGeocodingResultsBounds[1]); // upper right
            }
            let atLeastOneMarkerInMaxBounds = false;
            const markerBounds = new google.maps.LatLngBounds();
            const boundsForMarkersInMaxBounds = new google.maps.LatLngBounds();
            props.markers.forEach((marker) => {
                const position = {
                    lat: marker.position.geometry.coordinates[1],
                    lng: marker.position.geometry.coordinates[0]
                };
                markerBounds.extend(position);

                if (!maxMapBounds.isEmpty() && maxMapBounds.contains(position)) {
                    atLeastOneMarkerInMaxBounds = true;
                    boundsForMarkersInMaxBounds.extend(position);
                }
            });

            const bounds = atLeastOneMarkerInMaxBounds ? boundsForMarkersInMaxBounds : markerBounds;

            // hack for single markers: see https://stackoverflow.com/questions/3334729/google-maps-v3-fitbounds-zoom-too-close-for-single-marker
            if (bounds.getNorthEast().equals(bounds.getSouthWest())) {
                const extendPoint1 = new google.maps.LatLng(
                    bounds.getNorthEast().lat() + 0.001,
                    bounds.getNorthEast().lng() + 0.001
                );
                const extendPoint2 = new google.maps.LatLng(
                    bounds.getNorthEast().lat() - 0.001,
                    bounds.getNorthEast().lng() - 0.001
                );
                bounds.extend(extendPoint1);
                bounds.extend(extendPoint2);
            }
            map.fitBounds(bounds);
        }
    }, [props.shouldFitBounds]);

    React.useEffect(() => {
        if (!placesInfoWindow) {
            return;
        }
        placesInfoWindow.close();
        if (props.infoWindow) {
            placesInfoWindow.setContent(props.infoWindow.content);
            placesInfoWindow.setPosition({
                lat: props.infoWindow.position.geometry.coordinates[1],
                lng: props.infoWindow.position.geometry.coordinates[0]
            });
            placesInfoWindow.open(map);
        }
    }, [props.infoWindow]);

    return isLoaded ? (
        <GoogleMap
            mapContainerStyle={{
                boxSizing: 'border-box' as const,
                position: 'relative' as const,
                width: '100%',
                height: props.height || '40rem',
                border: '1px solid rgba(0,0,0,0.2)'
            }}
            center={center}
            zoom={props.defaultZoom || 10}
            onLoad={onLoad}
            onUnmount={onUnmount}
            onClick={(e) => onPositionChange(e, 'mapClicked')}
            onZoomChanged={onZoomChange}
            onBoundsChanged={onBoundsChanged}
            clickableIcons={false}
        >
            {props.markers.map((markerData, index) => (
                <Marker
                    key={`marker${index}${markerData.position.geometry.coordinates[0]}_${markerData.position.geometry.coordinates[1]}`}
                    position={toLatLng(markerData.position)}
                    onDragEnd={markerData.draggable ? (e) => onPositionChange(e, 'markerDragged') : undefined}
                    draggable={markerData.draggable}
                    icon={{
                        url: markerData.icon.url,
                        size: new google.maps.Size(markerData.icon.size[0], markerData.icon.size[1]),
                        scaledSize: new google.maps.Size(markerData.icon.size[0], markerData.icon.size[1])
                    }}
                    onClick={markerData.onClick}
                />
            ))}
        </GoogleMap>
    ) : (
        <InputLoading />
    );
};

export default InputMapGoogle;
