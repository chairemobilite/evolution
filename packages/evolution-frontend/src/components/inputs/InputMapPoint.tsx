/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import GeoJSON from 'geojson';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faMapMarkerAlt } from '@fortawesome/free-solid-svg-icons/faMapMarkerAlt';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import projectConfig from 'chaire-lib-common/lib/config/shared/project.config';
import { InputMapPointType } from 'evolution-common/lib/services/widgets';
import * as surveyHelper from 'evolution-common/lib/utils/helpers';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { CliUser } from 'chaire-lib-common/lib/services/user/userType';
import { WithTranslation, withTranslation } from 'react-i18next';
import { FeatureGeocodedProperties, MarkerData } from './maps/InputMapTypes';

// TODO Allow to support multiple maps and geocoders
import InputMapGoogle from './maps/google/InputMapGoogle';
import { geocodeSinglePoint } from './maps/google/GoogleGeocoder';

export interface InputMapPointProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    id: string;
    onValueChange: (e: any) => void;
    value?: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties>;
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: CliUser;
    inputRef?: React.LegacyRef<HTMLInputElement>;
    widgetConfig: InputMapPointType<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
}

interface InputMapPointState {
    defaultCenter: any;
    currentBounds?: [number, number, number, number];
    markers: MarkerData[];
}

/**
 * Allow to select a single point from a map.
 *
 * TODO For now, it only uses google map, but this class should remain map
 * agnostic and support more map types
 */
export class InputMapPoint<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> extends React.Component<
    InputMapPointProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation,
    InputMapPointState
> {
    private geocodeButtonRef: React.RefObject<HTMLButtonElement> = React.createRef();
    private iconUrl: string;

    constructor(props: InputMapPointProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation) {
        super(props);

        // get initial map center from current value or widget config (can be a function)
        const defaultCenter = props.value
            ? { lon: props.value.geometry.coordinates[0], lat: props.value.geometry.coordinates[1] }
            : surveyHelper.parse(props.widgetConfig.defaultCenter, props.interview, props.path, props.user) ||
              projectConfig.mapDefaultCenter;

        this.iconUrl = this.props.widgetConfig.icon
            ? surveyHelper.parseString(
                this.props.widgetConfig.icon.url,
                this.props.interview,
                this.props.path,
                this.props.user
            ) || '/dist/images/default_marker.svg'
            : '/dist/images/default_marker.svg';

        this.state = {
            defaultCenter,
            currentBounds: undefined,
            markers: this.props.value
                ? [
                    {
                        position: this.props.value,
                        icon: { url: this.iconUrl },
                        draggable: true
                    }
                ]
                : []
        };
    }

    onValueChange = (feature?: GeoJSON.Feature<GeoJSON.Point, FeatureGeocodedProperties>) => {
        this.props.onValueChange({ target: { value: feature } });
        this.setState({
            defaultCenter: feature
                ? { lat: feature.geometry.coordinates[1], lon: feature.geometry.coordinates[0] }
                : this.state.defaultCenter,
            markers: feature
                ? [
                    {
                        position: feature,
                        icon: { url: this.iconUrl },
                        draggable: true
                    }
                ]
                : []
        });
    };

    onMapReady = (bbox?: [number, number, number, number]) => {
        this.setState({ currentBounds: bbox });
        if (!this.props.value) {
            this.geocodeAddress(bbox);
        }
    };

    onBoundsChanged = (bbox?: [number, number, number, number]) => {
        this.setState({ currentBounds: bbox });
    };

    onGeocodeAddress = (e: React.MouseEvent) => {
        e.preventDefault();
        this.geocodeAddress(this.state.currentBounds);
    };

    getGeocodingQueryString = () => {
        return this.props.widgetConfig.geocodingQueryString
            ? this.props.widgetConfig.geocodingQueryString(this.props.interview, this.props.path, this.props.user)
            : undefined;
    };

    geocodeAddress = async (bbox?: [number, number, number, number]) => {
        const geocodingQueryString = this.getGeocodingQueryString();
        if (geocodingQueryString) {
            try {
                const feature = await geocodeSinglePoint(geocodingQueryString, { bbox });
                this.onValueChange(feature);
            } catch (error) {
                this.onValueChange(undefined);
            }
        }
    };

    render() {
        const maxZoom = this.props.widgetConfig.maxZoom || 18;

        return (
            <div className={'survey-question__input-map-container'}>
                <input // we don't really need the input with maps. We just need to find a way to change style of map container according to isEmpty or isValid status
                    type="hidden"
                    className={'apptr__form-input input-map-hidden'}
                    name={this.props.widgetConfig.shortname}
                    id={this.props.id}
                    defaultValue={JSON.stringify(this.props.value)}
                    readOnly={true}
                    ref={this.props.inputRef}
                />
                {this.props.widgetConfig.refreshGeocodingLabel && (
                    <button
                        type="button"
                        id={`${this.props.id}_refresh`}
                        className="button refresh-geocode green large"
                        onClick={this.onGeocodeAddress}
                        ref={this.geocodeButtonRef}
                    >
                        <FontAwesomeIcon icon={faMapMarkerAlt} className="faIconLeft" />
                        {surveyHelper.translateString(
                            this.props.widgetConfig.refreshGeocodingLabel,
                            this.props.i18n,
                            this.props.interview,
                            this.props.path,
                            this.props.user
                        )}
                    </button>
                )}
                <div aria-hidden="true" className="survey-question__input-map-container">
                    <InputMapGoogle
                        defaultCenter={this.state.defaultCenter}
                        value={this.props.value}
                        onValueChange={this.onValueChange}
                        defaultZoom={Math.min(this.props.widgetConfig.defaultZoom || 16, maxZoom)}
                        markers={this.state.markers}
                        onMapReady={this.onMapReady}
                        onBoundsChanged={this.onBoundsChanged}
                    />
                </div>
            </div>
        );
    }
}

export default withTranslation()(InputMapPoint);
