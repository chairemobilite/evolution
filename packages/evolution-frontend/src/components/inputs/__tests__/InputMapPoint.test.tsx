/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import InputMapPoint from '../InputMapPoint';
import { geocodeSinglePoint } from '../maps/google/GoogleGeocoder';

// Mock db queries
jest.mock('../maps/google/GoogleGeocoder', () => ({
    geocodeSinglePoint: jest.fn()
}));
const mockedGeocode = geocodeSinglePoint as jest.MockedFunction<typeof geocodeSinglePoint>;

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

const interviewAttributes: UserInterviewAttributes = {
    id: 1,
    uuid: 'arbitrary uuid',
    user_id: 1,
    is_completed: false,
    responses: {
        section1: {
            q1: 'abc',
            q2: 3
        },
        section2: {
            q1: 'test'
        }
    },
    validations: {
        section1: {
            q1: true,
            q2: false
        },
        section2: {
            q1: true
        }
    },
    is_valid: true
};

const baseWidgetConfig = {
    type: 'question' as const,
    twoColumns: true,
    path: 'test.foo',
    containsHtml: true,
    label: {
        fr: `Texte en français`,
        en: `English text`
    },
    inputType: 'mapPoint' as const,
    defaultCenter: { lat: 45, lon: -73 }
};

describe('Render InputMapPoint with various parameters', () => {

    test('Test with minimal parameters', () => {
        const wrapper = TestRenderer.create(
            <InputMapPoint
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={baseWidgetConfig}
                value={undefined}
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Test with all parameters', () => {
            
        const testWidgetConfig = Object.assign({
            geocodingQueryString: jest.fn(),
            refreshGeocodingLabel: {
                fr: `Rafraîchir la carte`,
                en: `Refresh map`
            },
            icon: {
                url: 'path/to/icon'
            },
            maxZoom: 18,
            defaultZoom: 15
        }, baseWidgetConfig);
        const wrapper = TestRenderer.create(
            <InputMapPoint
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={testWidgetConfig}
                value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
                inputRef={React.createRef()}
                size='medium'
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
    
});

describe('Test geocoding requests', () => {
    const testId = 'test';
    const geocodingString = '123 foo street';
    const mockOnValueChange = jest.fn();

    const testWidgetConfig = Object.assign({
        geocodingQueryString: jest.fn().mockReturnValue(geocodingString),
        refreshGeocodingLabel: {
            fr: `Geocode`,
            en: `Geocode`
        },
        icon: {
            url: 'path/to/icon'
        },
        maxZoom: 18,
        defaultZoom: 15,
        placesIcon: {
            url: 'path/to/icon'
        },
        updateDefaultValueWhenResponded: true
    }, baseWidgetConfig);

    const geocodedFeature = { 
        type: 'Feature' as const, 
        geometry: { type: 'Point' as const, coordinates: [-73.2, 45.1] },
        properties: { geocodingResultMetadata: { formattedAddress: '123 foo street, Montreal, QC' }, lastAction: 'geocoding', geocodingQueryString: geocodingString }
    };
    

    beforeEach(() => {
        mockedGeocode.mockClear();
        mockOnValueChange.mockClear();
    });

    test('Geocode single result', async () => {
    
        const mapPointWidget = mount(<InputMapPoint
            id={testId}
            onValueChange={mockOnValueChange}
            widgetConfig={testWidgetConfig}
            value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
            inputRef={React.createRef()}
            size='medium'
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
        />);

        // Find and click on the Geocode button, to return multiple values
        mockedGeocode.mockResolvedValueOnce(geocodedFeature);
        const geocodeButton = mapPointWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('click');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // The select list should display 4 children (2 places and the 2 extra elements) and should not have a confirm button
        mapPointWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        
        // Make sure the value has not been changed
        expect(mockOnValueChange).toHaveBeenCalledTimes(1);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: geocodedFeature }});
        
    });

    test('Geocode with undefined', async () => {
    
        const mapPointWidget = mount(<InputMapPoint
            id={testId}
            onValueChange={mockOnValueChange}
            widgetConfig={testWidgetConfig}
            value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
            inputRef={React.createRef()}
            size='medium'
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
        />);

        // Find and click on the Geocode button, to return multiple values
        mockedGeocode.mockResolvedValueOnce(undefined);
        const geocodeButton = mapPointWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('click');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // The select list should display 4 children (2 places and the 2 extra elements) and should not have a confirm button
        mapPointWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        
        // Make sure the value has not been changed
        expect(mockOnValueChange).toHaveBeenCalledTimes(1);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: undefined }});
        
    });

    test('Geocode with rejection', async () => {
    
        const mapPointWidget = mount(<InputMapPoint
            id={testId}
            onValueChange={mockOnValueChange}
            widgetConfig={testWidgetConfig}
            value={{ type: 'Feature' as const, properties: {}, geometry: { type: 'Point' as const, coordinates: [-73.1, 45.02]}}}
            inputRef={React.createRef()}
            size='medium'
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
        />);

        // Find and click on the Geocode button, to return multiple values
        mockedGeocode.mockRejectedValueOnce('error geocoding');
        const geocodeButton = mapPointWidget.find({id: `${testId}_refresh`, type: 'button'});
        expect(geocodeButton.getDOMNode<HTMLButtonElement>().textContent).toBe('Geocode');
        geocodeButton.simulate('click');
        // Let async functions terminate
        await new Promise(process.nextTick);
    
        // The select list should display 4 children (2 places and the 2 extra elements) and should not have a confirm button
        mapPointWidget.update();
        expect(mockedGeocode).toHaveBeenCalledTimes(1);
        expect(mockedGeocode).toHaveBeenCalledWith(geocodingString, expect.anything());
        
        // Make sure the value has not been changed
        expect(mockOnValueChange).toHaveBeenCalledTimes(1);
        expect(mockOnValueChange).toHaveBeenCalledWith({ target: { value: undefined }});
        
    });

});
