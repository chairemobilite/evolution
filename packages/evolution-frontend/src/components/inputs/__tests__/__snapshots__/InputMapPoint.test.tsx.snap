// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render InputMapPoint with various parameters Test with all parameters 1`] = `
<div
  className="survey-question__input-map-container"
>
  <input
    className="apptr__form-input input-map-hidden"
    defaultValue="{"type":"Feature","properties":{},"geometry":{"type":"Point","coordinates":[-73.1,45.02]}}"
    id="test"
    readOnly={true}
    type="hidden"
  />
  <button
    className="button refresh-geocode green large"
    id="test_refresh"
    onClick={[Function]}
    style={
      {
        "display": "none",
      }
    }
    type="button"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-location-dot faIconLeft"
      data-icon="location-dot"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={{}}
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M215.7 499.2C267 435 384 279.4 384 192C384 86 298 0 192 0S0 86 0 192c0 87.4 117 243 168.3 307.2c12.3 15.3 35.1 15.3 47.4 0zM192 128a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"
        fill="currentColor"
        style={{}}
      />
    </svg>
    Refresh map
  </button>
  <div
    aria-hidden="true"
    className="survey-question__input-map-container"
  >
    <div
      className="loader"
    >
      <p>
        Loading
        ...
      </p>
    </div>
  </div>
</div>
`;

exports[`Render InputMapPoint with various parameters Test with minimal parameters 1`] = `
<div
  className="survey-question__input-map-container"
>
  <input
    className="apptr__form-input input-map-hidden"
    id="test"
    readOnly={true}
    type="hidden"
  />
  <div
    aria-hidden="true"
    className="survey-question__input-map-container"
  >
    <div
      className="loader"
    >
      <p>
        Loading
        ...
      </p>
    </div>
  </div>
</div>
`;
